core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -182,43 +184,68 @@ public boolean init(Endpoint endpoint) {
             }
 
             return true;
-        } catch (Exception e) {
+        } catch (Throwable t) {
             if (eventLoop == null) {
                 // Always set the eventLoop because it can be used in a decorator.
                 eventLoop = CommonPools.workerGroup().next();
             }
-            failEarly(e);
+            failEarly(t);
         }
 
         return false;
     }
 
+    private void updateEndpoint(Endpoint endpoint) {
+        this.endpoint = endpoint;
+        autoFillSchemeAndAuthority();
+    }
+
     private void runThreadLocalContextCustomizer() {
         final Consumer<ClientRequestContext> customizer = THREAD_LOCAL_CONTEXT_CUSTOMIZER.get();
         if (customizer != null) {
             customizer.accept(this);
         }
     }
 
-    private void failEarly(Exception cause) {
+    private void failEarly(Throwable cause) {
         final RequestLogBuilder logBuilder = logBuilder();
         final UnprocessedRequestException wrapped = new UnprocessedRequestException(cause);
         logBuilder.endRequest(wrapped);
         logBuilder.endResponse(wrapped);
 
         final Request req = request();
         if (req instanceof StreamMessage) {
+            autoFillSchemeAndAuthority();
             ((StreamMessage<?>) req).abort();
         }
     }
 
+    private void autoFillSchemeAndAuthority() {
+        final Request req = request();
+        if (req instanceof HttpRequest) {
+            final HttpRequest httpReq = (HttpRequest) req;
+            final RequestHeaders headers = httpReq.headers();
+            final String authority = endpoint != null ? endpoint.authority() : "UNKNOWN";
+            if (headers.scheme() != null && authority.equals(headers.authority())) {
+                unsafeUpdateRequest(req);
```

Comment:
It seems like the `req` is not changed. Do we need this?

Reply:
No, we don't. Fixed. :sweat_smile: